biggus/__init__.py

```diff
@@ -561,6 +563,175 @@ def _normalise_keys(self, keys):
+    def __init__(self, contained_array):
+        self.array = contained_array
+
+    def __repr__(self):
+        return 'ArrayContainer({!r})'.format(self.array)
+
+    @property
+    def ndim(self):
+        return self.array.ndim
+
+    @property
+    def dtype(self):
+        return self.array.dtype
+
+    @property
+    def shape(self):
+        return self.array.shape
+
+    def __getitem__(self, keys):
+        # If this array handles np.newaxis, so too will this container.
+        return self.array[keys]
+
+    def ndarray(self):
+        try:
+            return self.array.ndarray()
+        except AttributeError:
+            return np.array(self.array)
+
+    def masked_array(self):
+        try:
+            return self.array.masked_array()
+        except AttributeError:
+            return np.ma.masked_array(self.array)
+
+
+class NewAxesArray(ArrayContainer):
+    def __init__(self, array, new_axes):
+        """
+        Creates an array which has new axes (i.e. length 1) at the
+        specified locations.
+
+        Parameters
+        ----------
+        array - array like
+            The array upon which to put new axes
+        new_axes - iterable of length array.ndim + 1
+            The number of new axes for each axis.
+            e.g. [2, 1, 0] for a 2d array gain two new axes on the left hand
+            side, one in the middle, and 0 on the right hand side.
+
+        """
+        super(NewAxesArray, self).__init__(array)
+        if self.array.ndim + 1 != len(new_axes):
+            raise ValueError('The new_axes must have length {} but was '
+                             'actually length {}.'.format(self.array.ndim + 1,
+                                                          len(new_axes)))
+        new_axes = np.array(new_axes)
+        dtype_kind = new_axes.dtype.type
+        if (not issubclass(dtype_kind, np.integer) or np.any(new_axes < 0)):
+            raise ValueError('Only positive integer types may be used for '
+                             'new_axes.')
+        self._new_axes = new_axes
+
+    @property
+    def ndim(self):
+        return np.sum(self._new_axes) + self.array.ndim
+
+    @property
+    def shape(self):
+        shape = list(self.array.shape)
+        # Starting from the higher dimensions, insert 1s at the locations
+        # of new axes.
+        for axes, n_new_axes in reversed(list(enumerate(self._new_axes))):
+            for _ in range(n_new_axes):
+                shape.insert(axes, 1)
+        return tuple(shape)
+
+    def _newaxis_keys(self):
+        # Compute the keys needed to produce an array of appropriate newaxis.
+        keys = [slice(None)] * self.array.ndim
+        # Starting from the higher dimensions, insert 1s at the locations
+        # of new axes.
+        for axes, n_new_axes in reversed(list(enumerate(self._new_axes))):
+            for _ in range(n_new_axes):
+                keys.insert(axes, np.newaxis)
```

Comment:
I find my form more readable, though I accept it is probably less efficient. I'm on the fence, push me one way or the other.